designer/server/src/main/scala/pl/touk/nussknacker/ui/limits/LimitsService.scala

```diff
@@ -0,0 +1,168 @@
+    }
+  }
+
+  private def checkAllLimits(
+      deployingScenario: ProcessName,
+      deployingScenarioProcessingType: ProcessingType,
+      deploymentUpdateStrategy: DeploymentUpdateStrategy,
+  )(implicit user: LoggedUser) = {
+    for {
+      _ <- checkPerProcessingTypeLimits(deployingScenario, deployingScenarioProcessingType, deploymentUpdateStrategy)
+      _ <- checkGlobalLimits(deployingScenario, deploymentUpdateStrategy)
+    } yield ()
+  }
+
+  private def limitsServiceLock = {
+    // note: currently we use in memory lock. It exposes Resource[Io, Unit] interface. It can be easily implemented
+    //       e.g. as a Postgresql-based lock. We may want to do this when the want to have Designer deployed in HA
+    LimitsService.mutex.lock
+  }
+
+  private def checkPerProcessingTypeLimits(
+      deployingScenario: ProcessName,
+      deployingScenarioProcessingType: ProcessingType,
+      deploymentUpdateStrategy: DeploymentUpdateStrategy,
+  )(implicit user: LoggedUser): EitherT[IO, MaxActiveScenariosCountExceededError, Unit] = {
+    perProcessingTypesLimitsProvider.forProcessingType(deployingScenarioProcessingType) match {
+      case Some(LimitsConfig(Some(maxActiveScenariosCount))) =>
+        EitherT {
+          getActiveScenariosFor(deployingScenarioProcessingType :: Nil)
+            .map { currentlyActiveScenarios =>
+              checkCurrentlyActiveScenariosCount(
+                deployingScenario,
+                currentlyActiveScenarios,
+                maxActiveScenariosCount.value,
+                deploymentUpdateStrategy
+              )
+            }
+        }
+      case Some(LimitsConfig(None)) | None =>
+        EitherT.right(IO.unit)
+    }
+  }
+
+  private def checkGlobalLimits(
+      deployingScenario: ProcessName,
+      deploymentUpdateStrategy: DeploymentUpdateStrategy,
+  ): EitherT[IO, MaxActiveScenariosCountExceededError, Unit] = {
+    globalLimitsConfig.maxActiveScenariosCount match {
+      case Some(maxActiveScenariosCount) =>
+        EitherT {
+          implicit val user: LoggedUser = NussknackerInternalUser.instance
+          val allProcessingTypes        = perProcessingTypesLimitsProvider.all.keys
+          getActiveScenariosFor(allProcessingTypes)
+            .map { currentlyActiveScenarios =>
+              checkCurrentlyActiveScenariosCount(
+                deployingScenario,
+                currentlyActiveScenarios,
+                maxActiveScenariosCount.value,
+                deploymentUpdateStrategy
+              )
+            }
+        }
+      case None =>
+        EitherT.right(IO.unit)
+    }
+  }
+
+  private def checkCurrentlyActiveScenariosCount(
+      currentlyDeployingScenario: ProcessName,
+      currentlyActiveScenarios: Set[ProcessName],
+      maxActiveScenariosCount: Int,
+      deploymentUpdateStrategy: DeploymentUpdateStrategy,
+  ) = {
+    val maxActiveScenariosIncludingCurrentlyDeployingScenario = deploymentUpdateStrategy match {
```

Comment:
Please don't modify both left and right sides of an expression.
This expression is as simple as:
```
val activeScenariosCountAfterDeployment = deploymentUpdateStrategy match {
// case for redeploy => currentlyActiveScenarios.size
// case for normal deploy  => currentlyActiveScenarios.size + 1
}
activeScenariosCountAfterDeployment <= maxActiveScenariosCount
```